src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -1445,10 +1446,36 @@ int RGWBucketAdminOp::remove_bucket(rgw::sal::Driver* driver, RGWBucketAdminOpSt
   if (ret < 0)
     return ret;
 
+  // check if the bucket is owned by my zonegroup, if not, refuse to remove it
+  if (!forwarded_request && bucket->get_info().zonegroup != driver->get_zone()->get_zonegroup().get_id()) {
+    ldpp_dout(dpp, 0) << "ERROR: The bucket's zonegroup does not match. "
+                      << "Removal is not allowed. Please execute this operation "
+                      << "on the bucket's zonegroup: " << bucket->get_info().zonegroup << dendl;
+    return -ERR_PERMANENT_REDIRECT;
+  }
+
   if (bypass_gc)
     ret = bucket->remove_bypass_gc(op_state.get_max_aio(), keep_index_consistent, y, dpp);
   else
     ret = bucket->remove(dpp, op_state.will_delete_children(), y);
+  if (ret < 0)
+    return ret;
+
+  // forward to master zonegroup
+  const std::string delpath = "/" + dpp->get_cct()->_conf->rgw_admin_entry + "/bucket";
+  RGWEnv env;
+  env.set("REQUEST_METHOD", "DELETE");
+  env.set("SCRIPT_URI", delpath);
+  env.set("REQUEST_URI", delpath);
+  env.set("QUERY_STRING", fmt::format("bucket={}&tenant={}", bucket->get_name(), bucket->get_tenant()));
+  req_info req(dpp->get_cct(), &env);
+
+  ret = rgw_forward_request_to_master(dpp, site, std::nullopt, nullptr, nullptr, req, y);
```

Comment:
could we pass `bucket->get_owner()` here to avoid needing optional support from rgw_forward_request_to_master?

Reply:
no objection, just wouldn't thi cause any possible confusion of `rgwx-uid` usage? to me, it might be treated differently per API on how we utilise that, but so just asking to make sure.
Cause here we want to represent a real system request that doesn't have any other ACLOwner that itself.

Reply:
i don't think the `rgwx-uid` will matter for the `DELETE /admin/bucket` api

this is just a strange use of `rgw_forward_request_to_master()`. here we're constructing a new http request to send, instead of forwarding one with an existing identity. maybe there's some function other than `RGWRESTConn::forward()` that we should be calling instead, but then we'd have to duplicate most of `rgw_forward_request_to_master()` to use that

Reply:
agreed. I've changed it in the new commit.